Removed ThisShouldNotHappenError

lutovich opened the pull request:
Usages replaced by either `IllegalStateException`, `IllegalArgumentException` or `UnsupportedOperationException` depending on the use case.

systay commented:
Nice stuff!

chrisvest commented on community/kernel/src/main/java/org/neo4j/helpers/ThisShouldNotHappenError.java:
```diff
@@ -1,34 +0,0 @@
- */
-package org.neo4j.helpers;
-
-public class ThisShouldNotHappenError extends Error
```
You can't remove this class, because it has to be deprecated first.

chrisvest commented:
Many of these exceptions end up with rather odd exception messages. I also think that many of them should be `AssertionErrors`.

lutovich committed:
Removed usages of ThisShouldNotHappenError

Usages replaced by either IllegalStateException, IllegalArgumentException,
UnsupportedOperationException or AssertionError depending on the use case.
The class itself can't be removed right now because we have a binary dependency
on the previous cypher-compiler which uses this exception.

lutovich commented:
@chrisvest changed some places to use `AssertionError`. `ThisShouldNotHappenError` itself can't indeed be removed now because 2.3 cypher compiler depends on it. So now it is just deprecated.
